File: Optifi/test/utils/pda.ts
import * as anchor from "@project-serum/anchor";
import Config from "../config.json";
import {
  AccountLayout,
  MintLayout,
  TOKEN_PROGRAM_ID,
  Token,
} from "@solana/spl-token";
import {
  PublicKey,
  SystemProgram,
  Transaction,
  Keypair,
  SYSVAR_RENT_PUBKEY,
} from "@solana/web3.js";

const programId = new PublicKey(Config.OptiFi_ProgramId);
const serumDexProgramId = new PublicKey(Config.Serum_Dex_ProgramId); // serum dex program id on devnet

// ==================================================================
// For Optifi Exchange use
// ==================================================================
export const getOptiFiExchange = async (uuid: String) => {
  return await anchor.web3.PublicKey.findProgramAddress(
    [Buffer.from("optifi_exchange"), Buffer.from(uuid)],
    programId
  );
};
/// PDA is the account which controls all user's usdc vaults
export const getPDA = async (optifi_exchange: PublicKey) => {
  return await anchor.web3.PublicKey.findProgramAddress(
    [Buffer.from("user_token_account_pda"), optifi_exchange.toBuffer()],
    programId
  );
};

export const getCentralUsdcPoolAuth = async (
  optifi_exchange_id: PublicKey = new PublicKey(Config.OptiFi_Exchange_Id)
) => {
  return await anchor.web3.PublicKey.findProgramAddress(
    [Buffer.from("central_usdc_pool_auth"), optifi_exchange_id.toBuffer()],
    programId
  );
};

// ==================================================================
// For User Account use
// ==================================================================

// user account(pda) controls all spl token vaults of the user
export const getUserAccount = async (
  optifi_exchange_id: PublicKey,
  wallet: PublicKey
) => {
  return await anchor.web3.PublicKey.findProgramAddress(
    [
      Buffer.from("user_account"),
      optifi_exchange_id.toBuffer(),
      wallet.toBuffer(),
    ],
    programId
  );
};

// ==================================================================
// For Serum Orderbook use
// ==================================================================
const getVaultOwnerAndNonce = async (
  marketPublicKey: PublicKey,
  dexProgramId = serumDexProgramId
) => {
  return await anchor.web3.PublicKey.findProgramAddress(
    [marketPublicKey.toBuffer()],
    dexProgramId
  );
};

export const getVaultOwnerAndNonce2 = async (
  marketPublicKey: PublicKey,
  dexProgramId = serumDexProgramId
): Promise<[anchor.web3.PublicKey, anchor.BN]> => {
  const nonce = new anchor.BN(0);
  while (nonce.toNumber() < 255) {
    try {
      const vaultOwner = await PublicKey.createProgramAddress(
        [marketPublicKey.toBuffer(), nonce.toArrayLike(Buffer, "le", 8)],
        dexProgramId
      );
      return [vaultOwner, nonce];
    } catch (e) {
      nonce.iaddn(1);
    }
  }
  throw new Error("Unable to find nonce");
};

// optifi_market_mint_auth pda is the account which can mint instrument spl tokens to user
export const get_optifi_market_mint_auth_pda = async (
  optifi_exchange: PublicKey = new PublicKey(Config.OptiFi_Exchange_Id)
) => {
  return await anchor.web3.PublicKey.findProgramAddress(
    [Buffer.from("optifi_market_mint_auth"), optifi_exchange.toBuffer()],
    programId
  );
};

// optifi_market_auth pda is the account which can has the authority over the creating serum market
export const getSerumMarketAuthority = async (
  optifi_exchange: PublicKey = new PublicKey(Config.OptiFi_Exchange_Id)
) => {
  return await anchor.web3.PublicKey.findProgramAddress(
    [Buffer.from("serum_market_auth"), optifi_exchange.toBuffer()],
    programId
  );
};

// optifi_prune_auth pda is the account which can has the prune authority
// over the open orders accounts on the creating serum market
export const getSerumPruneAuthority = async (
  optifi_exchange: PublicKey = new PublicKey(Config.OptiFi_Exchange_Id)
) => {
  return await anchor.web3.PublicKey.findProgramAddress(
    [Buffer.from("serum_prune_auth"), optifi_exchange.toBuffer()],
    programId
  );
};

// ==================================================================
// For Optifi Market use
// ==================================================================
export const getOptifiMarket = async (
  exchangePk: PublicKey,
  marketId: number
) => {
  return await anchor.web3.PublicKey.findProgramAddress(
    [
      Buffer.from("optifi_market"),
      exchangePk.toBuffer(),
      // Uint8Array.of(id),
      new anchor.BN(marketId).toArrayLike(Buffer, "be", 8),
    ],
    programId
  );
};

export const getDexOpenOrders = async (
  optifi_exchange_id: PublicKey = new PublicKey(Config.OptiFi_Exchange_Id),
  serumMarketId: PublicKey = new PublicKey(
    Config.Created_Serum_Orderbook.market
  ),
  user_account: PublicKey
) => {
  return await anchor.web3.PublicKey.findProgramAddress(
    [
      Buffer.from("serum_open_orders"),
      optifi_exchange_id.toBuffer(),
      serumMarketId.toBuffer(),
      user_account.toBuffer(),
    ],
    programId
  );
};

// ==================================================================
// For AMM use
// ==================================================================

// amm account(pda) controls spl tokens execept usdc vaults of the Amm
export const getAmmAccount = async (
  optifi_exchange_id: PublicKey,
  amm_idx: number
) => {
  return await anchor.web3.PublicKey.findProgramAddress(
    [
      Buffer.from("amm"),
      optifi_exchange_id.toBuffer(),
      // Buffer.from([amm_idx])
      Uint8Array.of(amm_idx),
    ],
    programId
  );
};

// amm usdc vault auth(pda) controls the usdc vault of ALL AMMs
export const getAmmLiquidityAuth = async (
  optifi_exchange_id: PublicKey = new PublicKey(Config.OptiFi_Exchange_Id)
) => {
  return await anchor.web3.PublicKey.findProgramAddress(
    [Buffer.from("amm_liquidity_auth"), optifi_exchange_id.toBuffer()],
    programId
  );
};

File: Optifi/programs/Optifi/src/utils/pda.rs
use anchor_lang::prelude::*;

/// used to derive the address of an optifi exchange
pub const PREFIX_OPTIFI_EXCHANGE: &str = "optifi_exchange";

/// 1. used to derive the address of a user account
/// 2. this pda address has the authority over all spl token accounts and open orders accounts of the user
pub const PREFIX_USER_ACCOUNT: &str = "user_account";

/// used to derive the address of an instrument
pub const PREFIX_INSTRUMENT: &str = "instrument";

/// used to derive the address of an optifi market
pub const PREFIX_OPTIFI_MARKET: &str = "optifi_market";

/// 1. used to derive an address
/// 2. this pda has the mint authority over all instruments spl tokens for all optifi markets
pub const PREFIX_OPTIFI_MARKET_MINT_AUTH: &str = "optifi_market_mint_auth";

/// used to derive the address of a user's serum open orders
pub const PREFIX_SERUM_OPEN_ORDERS: &str = "serum_open_orders";

/// 1. used to derive an address
/// 2. this pda is the serum market authority and the serum market prune authority
pub const PREFIX_SERUM_MARKET_AUTH: &str = "serum_market_auth";

/// 1. used to derive address of the central usdc pool vault for fund settlement purpose
/// 2. this pda is the serum market authority and the serum market prune authority
pub const PREFIX_CENTRAL_USDC_POOL: &str = "central_usdc_pool";

/// 1. used to derive an address
/// 2. this pda has authority over central usdc pool vault
pub const PREFIX_CENTRAL_USDC_POOL_AUTH: &str = "central_usdc_pool_auth";

/// used to derive amm account address
pub const PREFIX_AMM: &str = "amm";

/// Used to derive market maker account address
pub const PREFIX_MARKET_MAKER: &str = "market_maker";

/// used to derive the authority address of amm's usdc vault,lp token mint and lp token vault
pub const PREFIX_AMM_LIQUIDITY_AUTH: &str = "amm_liquidity_auth";

/// used for the liquidation state account, which will record the progress of liquidations for a user
pub const PREFIX_LIQUIDATION_STATE: &str = "liquidation_state";

/// Used for market maker liquidity pool
pub const PREFIX_MM_LIQUIDITY_AUTH: &str = "mm_liquidity_auth";

/// get the user account pda and its bump seed
pub fn get_user_account_pda(
    optifi_exchange: &Pubkey,
    user: &Pubkey,
    program_id: &Pubkey,
) -> (Pubkey, u8) {
    Pubkey::find_program_address(
        &[
            PREFIX_USER_ACCOUNT.as_bytes(),
            optifi_exchange.as_ref(),
            user.as_ref(),
        ],
        program_id,
    )
}

/// get mint authority pda
pub fn get_optifi_market_mint_auth_pda(
    optifi_exchange: &Pubkey,
    program_id: &Pubkey,
) -> (Pubkey, u8) {
    Pubkey::find_program_address(
        &[
            PREFIX_OPTIFI_MARKET_MINT_AUTH.as_bytes(),
            optifi_exchange.as_ref(),
        ],
        program_id,
    )
}

/// get serum market and prune authority pda
pub fn get_serum_market_auth_pda(optifi_exchange: &Pubkey, program_id: &Pubkey) -> (Pubkey, u8) {
    Pubkey::find_program_address(
        &[
            PREFIX_SERUM_MARKET_AUTH.as_bytes(),
            optifi_exchange.as_ref(),
        ],
        program_id,
    )
}

/// get the central usdc pool vault address(pda)
pub fn get_central_usdc_pool_pda(optifi_exchange: &Pubkey, program_id: &Pubkey) -> (Pubkey, u8) {
    Pubkey::find_program_address(
        &[
            PREFIX_CENTRAL_USDC_POOL.as_bytes(),
            optifi_exchange.as_ref(),
        ],
        program_id,
    )
}

/// get the authority address (pda) of central usdc pool
pub fn get_central_usdc_pool_auth_pda(
    optifi_exchange: &Pubkey,
    program_id: &Pubkey,
) -> (Pubkey, u8) {
    Pubkey::find_program_address(
        &[
            PREFIX_CENTRAL_USDC_POOL_AUTH.as_bytes(),
            optifi_exchange.as_ref(),
        ],
        program_id,
    )
}

pub fn get_market_maker_pool_auth_pda(
    optifi_exchange: &Pubkey,
    market_maker_address: &Pubkey,
    program_id: &Pubkey,
) -> (Pubkey, u8) {
    Pubkey::find_program_address(
        &[
            PREFIX_MM_LIQUIDITY_AUTH.as_bytes(),
            optifi_exchange.as_ref(),
            market_maker_address.as_ref(),
        ],
        program_id,
    )
}

/// get the amm address (pda) with index
pub fn get_amm_pda(optifi_exchange: &Pubkey, amm_idx: u8, program_id: &Pubkey) -> (Pubkey, u8) {
    Pubkey::find_program_address(
        &[PREFIX_AMM.as_bytes(), optifi_exchange.as_ref(), &[amm_idx]],
        program_id,
    )
}

/// get the authority address (pda) of amm usdc vault, lp token mint and vault
pub fn get_amm_liquidity_auth_pda(optifi_exchange: &Pubkey, program_id: &Pubkey) -> (Pubkey, u8) {
    Pubkey::find_program_address(
        &[
            PREFIX_AMM_LIQUIDITY_AUTH.as_bytes(),
            optifi_exchange.as_ref(),
        ],
        program_id,
    )
}
